### DISCUSSION TITLE
Soldier of Fortune (Dreamcast)

### DISCUSSION THREAD

**Discussion by snake-biscuits** [snake-biscuits/bsp_tool#117] [Completed]
### Notes
Support for this format was confirmed in https://github.com/snake-biscuits/bsp_tool/commit/c08a75b9aab37bd0b1bce873ba1bc5b258cb1e5c
Keeping the thread up for future reference

### Original Post
Trawling through a `.gdi` dump, there are enough instances of `"classname" "worldspawn"` to account for 121 `.bsp`s
However, I can't seem to find a way to split the `.img` holding this data
Might have to reverse engineer the executable to figure out the filesystem

I also haven't been able to locate `b"IBSP"` headers, so it's possible the format is unique

**Comment by snake-biscuits** [comment]
NVM I located 117 `b"IBSP", 46` headers
Will probably need to complete #21 to parse them all
Could potentially extract the headers manually & determine the filesize, but that's a lot of maps

**Reply by snake-biscuits** [reply]
Looks like there's some large recurring structures that interrupt file data
If they occur at regular intervals we might be able to circumvent that
I can find 122 `IBSP` entries on their own, so it seems like the recurring structure might interrupt some headers

This means #21 alone won't be enough, the `.gdi` `.img` segments still need to get reinterpreted into files somehow

**Reply by snake-biscuits** [reply]
The recurring structure is `130` bytes long, alignment doesn't seem constant
But maybe each one preceeds a constant number of bytes
Should probably do some research on `.gdi` structures

**Reply by snake-biscuits** [reply]
Might've found tools for browsing the GD-ROM filesystem:
multimedia.cx: [Understanding The Dreamcast GD-ROM Layout](https://multimedia.cx/eggs/understanding-the-dreamcast-gd-rom-layout/)
https://github.com/multimediamike/fuseiso3

**Reply by snake-biscuits** [reply]
Couldn't figure out fuseiso3 once I got it to compile
However, I found [IsoBuster](https://www.isobuster.com/) was able to extract `track06.bin` as a contiguous `.iso`
According to multimedia.cx the actual filesystem resides in `track03.bin`
I might be able to determine file sizes & names from there

**Comment by snake-biscuits** [comment]
I did also manage to locate some map name strings in the data (also some C headers?)
```
dm/nycctf1		"Gangland: 16"
dm/sibctf1		"Hell Freezes: 16"
dm/irqctf1		"Wastelands: 6 to 12"
dm/jpnctf1		"Parking: 4 to 8"
dm/sudctf1		"Flag Frag: 4 to 8"
dm/sudctf2		"Missile Base: 16"
dm/gerctf1		"Castle Kill: 6 to 10"
dm/nycdm1		"Warehouse: 8 to 16 players"
dm/nycdm2		"Underground: 4 to 6 players"
dm/nycdm3		"Pawn Shop : 2 to 4 players"
dm/nycdm4		"Back Alley: 2 to 4 players"
dm/nycdm5		"Bunker: 2 to 5 players"
dm/kosdm1		"The Sewers: 4 to 8 players"
dm/irqdm1		"Refinery: 6 to 8 players"
dm/irqdm2		"Palace: 8 to 10 players"
dm/jpndm2		"Restaurant: 2 to 6 players"
dm/raven		"Raven Software: 4+ players"
dm/suddm1		"WI Death Farm: 6 to 10 players"
dm/suddm2		"Trainyard: 6 to 8 players"
dm/suddm3		"Rusted: 2 to 6 players"
dm/nycdm3		"Pawn Shop : 2 to 4 players"
dm/nycdm4		"Back Alley: 2 to 4 players"
dm/nycdm5		"Bunker: 2 to 5 players"
dm/kosdm1		"The Sewers: 4 to 8 players"
dm/jpndm2		"Restaurant: 2 to 6 players"
dm/raven		"Raven Software: 4+ players"
dm/suddm1		"WI Death Farm: 6 to 10 players"
dm/suddm2		"Trainyard: 6 to 8 players"
```
Removed a recurring struct thing from the middle, a little confused by the repetition

**Reply by snake-biscuits** [reply]
Just found 105 `PRT1`[^prt] files
Got a sneaking suspicion Crave Entertainment might've shipped their source files

[^prt]: Valve Developer Community: [PRT](https://developer.valvesoftware.com/wiki/PRT)

**Reply by snake-biscuits** [reply]
The map name strings data looks like `pak3.pak/maplists/ravdm.lst` from my PC copy

**Reply by snake-biscuits** [reply]
The `.prt` files all seem to come just before `.bsp` data
Some `.eal` files seem to be mixed in too
My guess is the `maps/` folder is just all there in alphabetical order

I could probably determine `.prt` lengths by parsing them
Matching `.prt` order to `.bsp` order would be tricky though

**Comment by kungfooman** [comment]
> Got a sneaking suspicion Crave Entertainment might've shipped their source files

How does a portal file contain the map source? Do you mean the brushes of a map?

**Reply by snake-biscuits** [reply]
`.prt` files are a side-product of the VIS compile process (see #82)
You can find a few from my compiles in [`tests/mapsrc`](https://github.com/snake-biscuits/bsp_tool/tree/master/tests/mapsrc)

Finding a bunch of `.prt` data & C code makes me think Crave forgot to clean up before shipping
~~Haven't scoured the whole `GD-ROM` filesystem yet, but I might be able to find some `.map` source files
Assuming those really did get shipped~~

It's unlikely any uncompiled `.map` files will show up, since I scanned for `ENTITIES`
Entities are near identical in `.map` & `.bsp` form
I found `122` `IBSP` headers & `122` `worldspawn` strings, which means `122` `.bsp` & no `.map`

**Reply by kungfooman** [reply]
The C code is maybe like Quake 3 LCC bot code?

**Reply by snake-biscuits** [reply]
Just some `#define` integer constants, which could for anything, doesn't appear in my PC copy tho
Sample:
```c++
// if we wind up #define'ing more than 500 anims, update ANIMLISTSIZE in ds.exe/ds.cpp

#define SOLID_NOT	0
#define SOLID_SOLID 2  // erm, why wasn't this made SOLID_BBOX?  grrrr
#define SOLID_BSP	3
```
This could be for scripting actually
```c++
// template file for common directory

#include "../common/header.ds"

output "p:/base/ds/common"

parameter entity kill_thing
remove entity kill_thing // removes whatever was in parm1 
```

**Comment by snake-biscuits** [comment]
Extracted all `117` `.bsp`s with the following script:
```python
import struct

def substring_indices(file, sub: bytes, buf_max: int = 2048):
    # hopefully buffer alignment won't split the substring
    cur = 0
    while True:
        file.seek(cur)  # resume scanning
        buf = file.read(2048)
        cur = file.tell()
        if sub in buf:
            yield cur - 2048 + buf.index(sub)
        elif len(buf) == 0:
            break  # reached EOF

with open("Track 06.iso", "rb") as gd_file:
    for i, offset in enumerate(substring_indices(gd_file, b"IBSP\x2E\x00\x00\x00")):
        gd_file.seek(offset + 8)  # after ("IBSP", 46)
        # assuming we only have the same 19 headers as raven.soldier_of_fortune
        lump_headers = struct.iter_unpack("2I", gd_file.read(8 * 19))
        bsp_size = sum(sorted(lump_headers)[-1])
        gd_file.seek(offset)
        print(f"{i:03d} {offset:08X} writing to 'maps/entry_{i + 1:03d}.bsp' ...")
        with open(f"maps/entry_{i + 1:03d}.bsp", "wb") as bsp_file:
            bsp_file.write(gd_file.read(bsp_size))
```
An initial glance at the extracted maps with `bsp_tool` revealed a few things:
 * `bsp.loading_errors`: a good handful of lumps have changed
 * `bsp.ENTITIES[0].get("message", None)`: campaign maps have been split up
 * `min([h.offset for h in bsp.headers.values()])`: 3 new lump headers have been added

So `raven.soldier_of_fortune_dc` will have to be it's own branch

> It's worth noting that I can't confirm these files extracted correctly
> Getting the correct filenames & lengths will require proper `GD-ROM` extraction

**Reply by snake-biscuits** [reply]
> `bsp.loading_errors`: a good handful of lumps have changed
> So `raven.soldier_of_fortune_dc` will have to be it's own branch

Scratch that, I forgot `raven.soldier_of_fortune` overlaps w/ `id_software.quake2`
It's likely the map format hasn't changed besides lowering the engine limits
Loading with the correct branch creates no loading_errors

I will run the `min([h.offset for h in bsp.headers.values()])` test on PC `SoF` maps

**Reply by snake-biscuits** [reply]
`raven.soldier_of_fortune` lumps also begin at `184` bytes (22 headers)
I should remember this test for more `.bsp` formats
